app/js/services/auth.js

```diff
@@ -0,0 +1,82 @@
+import Cookies from 'js-cookie'
+import axios from 'axios'
+import { Promise } from 'es6-promise'
+import { CP_ADMIN_HOST, CP_ANALYTICS_HOST } from '../constants/app_constants'
+import { routeUtils } from '../helpers/router'
+
+const LOGIN_URL = CP_ADMIN_HOST + '#/auth/sign_in?return_url=' + CP_ANALYTICS_HOST
+const SESSION_PING_URL = CP_ADMIN_HOST + 'api/v1/session'
+const REFRESH_TOKEN_URL = CP_ADMIN_HOST + 'api/v1/api_session/refresh'
+
+const Auth = {
+  init () {
+    this.addXHRInterceptor()
+    this.initSession()
+  },
+
+  addXHRInterceptor () {
+    axios.interceptors.request.use(
+      this.addXHRHeaders.bind(this),
+      (error) => { return Promise.reject(error) }
+    )
+
+    axios.interceptors.response.use(
+      (response) => { return response },
+      (error) => { return this.resolveRejectedRequest(error) }
+    )
+  },
+
+  initSession () {
+    axios.get(SESSION_PING_URL)
+  },
+
+  addXHRHeaders (request) {
+    request.headers['Authorization'] = this.getAccessToken()
+    return request
+  },
+
+  resolveRejectedRequest (response) {
+    if (response.status === 401) {
+      return this.refreshToken(response).then((request) => {
+        // Restart previous request after token is refreshed
+        this.setAccessToken(request.data.access_token)
+        this.restartRequest(response)
+      }).catch(this.redirectToLoginPage)
+    } else {
+      return Promise.reject(response)
+    }
+  },
+
+  refreshToken (response) {
+    return axios.post(REFRESH_TOKEN_URL, {
+      refresh_token: this.getRefreshToken(),
+    })
+  },
+
+  redirectToLoginPage () {
+    routeUtils.redirectTo(LOGIN_URL)
+  },
+
+  getAccessToken () {
+    return Cookies.get('accessToken')
+  },
+
+  setAccessToken (newToken) {
+    Cookies.set('accessToken', newToken)
+  },
+
+  getRefreshToken () {
+    return Cookies.get('refreshToken')
+  },
+
+  restartRequest (response) {
+    return axios({
+      method: response.config.method,
+      url: response.config.url,
+      data: response.data,
+    })
+  },
+
+}
+
+export default Auth
```

Comment:
With ES6 this whole file can be simplified. As yesterday we were struggling with the tests I forgot about it. By exporting the `Auth` object we make available all the methods to anyone importing it from other files. If I am not wrong, the only method used from outside this file is `Auth.init()`. Then that is the only thing that we would need to export. And all the functions wrapped inside the `Auth` object can be functions that live in the root of the file, but not exported. By doing that they become private functions that you don't need to call with `this` anymore, neither you need to `.bind(this)`.

Something like this

```javascript
import Cookies from 'js-cookie'
import axios from 'axios'
import { Promise } from 'es6-promise'
import { CP_ADMIN_HOST, CP_ANALYTICS_HOST } from '../constants/app_constants'
import { routeUtils } from '../helpers/router'

const LOGIN_URL = CP_ADMIN_HOST + '#/auth/sign_in?return_url=' + CP_ANALYTICS_HOST
const SESSION_PING_URL = CP_ADMIN_HOST + 'api/v1/session'
const REFRESH_TOKEN_URL = CP_ADMIN_HOST + 'api/v1/api_session/refresh'

function addXHRInterceptor () {
  axios.interceptors.request.use((error) => Promise.reject(error))

  axios.interceptors.response.use(
    (response) => response
    (error) => this.resolveRejectedRequest(error)
  )
}

function initSession () {
  axios.get(SESSION_PING_URL)
}

function addXHRHeaders (request) {
  request.headers['Authorization'] = getAccessToken()
  return request
}

function resolveRejectedRequest (response) {
  if (response.status === 401) {
    return refreshToken(response).then((request) => {
      // Restart previous request after token is refreshed
      setAccessToken(request.data.access_token)
      restartRequest(response)
    }).catch(redirectToLoginPage)
  } else {
    return Promise.reject(response)
  }
}

function refreshToken (response) {
  return axios.post(REFRESH_TOKEN_URL, {
    refresh_token: getRefreshToken(),
  })
}

function redirectToLoginPage () {
  routeUtils.redirectTo(LOGIN_URL)
}

function getAccessToken () {
  return Cookies.get('accessToken')
}

function setAccessToken (newToken) {
  Cookies.set('accessToken', newToken)
}

function getRefreshToken () {
  return Cookies.get('refreshToken')
}

function restartRequest (response) {
  return axios({
    method: response.config.method,
    url: response.config.url,
    data: response.data,
  })
}

export const init = () => {
  addXHRInterceptor()
  initSession()
}
```

I exported the `init` method as a const and created the other as functions just to show that it works with both. I prefer having them as `const myFunctionName = () => {}` but just because it looks better from my point of view. As long as we are not using `this` the behaviour is the same. The only thing is that it should be consistent, either we use `const` as functions all the time or we use `function`.